annotated_test.go

```diff
@@ -86,6 +94,327 @@ func (a anotherStringer) String() string {
+						return &fromStringer{
+							name: myStringer.String(),
+						}
+					},
+					fx.From(new(anotherStringer)),
+				),
+			),
+			invoke: func(s fmt.Stringer) {
+				assert.Equal(t, s.String(), "another stringer")
+			},
+		},
+		{
+			desc: "provide with multiple types From",
+			provide: fx.Provide(
+				newFromStringer,
+				func() anotherStringer {
+					return anotherStringer{
+						"another stringer",
+					}
+				},
+				fx.Annotate(
+					func(myStringer1 myStringer, myStringer2 myStringer) fmt.Stringer {
+						return &fromStringer{
+							name: myStringer1.String() + " and " + myStringer2.String(),
+						}
+					},
+					fx.From(new(anotherStringer), new(*fromStringer)),
+				),
+			),
+			invoke: func(s fmt.Stringer) {
+				assert.Equal(t, s.String(), "another stringer and a good stringer")
+			},
+		},
+		{
+			desc: "provide from with param annotation",
+			provide: fx.Provide(
+				fx.Annotate(
+					newFromStringer,
+					fx.ResultTags(`name:"struct1"`),
+				),
+				fx.Annotate(
+					func(myStringer myStringer) fmt.Stringer {
+						return &fromStringer{
+							name: myStringer.String(),
+						}
+					},
+					fx.From(new(*fromStringer)),
+					fx.ParamTags(`name:"struct1"`),
+				),
+			),
+			invoke: func(s fmt.Stringer) {
+				assert.Equal(t, s.String(), "a good stringer")
+			},
+		},
+		{
+			// same as the test above, except now we annotate
+			// it in a different order.
+			desc: "provide from with param annotation, in different order",
+			provide: fx.Provide(
+				fx.Annotate(
+					newFromStringer,
+					fx.ResultTags(`name:"struct1"`),
+				),
+				fx.Annotate(
+					func(myStringer myStringer) fmt.Stringer {
+						return &fromStringer{
+							name: myStringer.String(),
+						}
+					},
+					fx.From(new(*fromStringer)),
+					fx.ParamTags(`name:"struct1"`),
+				),
+			),
+			invoke: func(s fmt.Stringer) {
+				assert.Equal(t, s.String(), "a good stringer")
+			},
+		},
+		{
+			desc: "annotate fewer items than required for constructor",
+			provide: fx.Provide(
+				newFromStringer,
+				func() anotherStringer {
+					return anotherStringer{
+						"another stringer",
+					}
+				},
+				fx.Annotate(
+					func(myStringer1 myStringer, fromStringer2 *fromStringer) fmt.Stringer {
+						return &fromStringer{
+							name: myStringer1.String() + " and " + fromStringer2.String(),
+						}
+					},
+					fx.From(new(anotherStringer)),
+				),
+			),
+			invoke: func(s fmt.Stringer) {
+				assert.Equal(t, s.String(), "another stringer and a good stringer")
+			},
+		},
+		{
+			desc: "Provide with empty From type",
+			provide: fx.Provide(
+				newFromStringer,
+				fx.Annotate(
+					func(myStringer *fromStringer) fmt.Stringer {
+						return &fromStringer{
+							name: myStringer.String(),
+						}
+					},
+					fx.From(),
+				),
+			),
+			invoke: func(s fmt.Stringer) {
+				assert.Equal(t, s.String(), "a good stringer")
+			},
+		},
+		{
+			desc: "Provide with variadic function",
```

Comment:
I'm wondering how From annotation will work when it's applied directly to the variadic parameter of the constructor. So something like this?
```
fx.Provide(
  fx.Annotate(
    func(a A, b ...B) C {...},
    fx.From(new(StructA), new(StructB)),
  ),
)
```
Could you investigate this? I'm guessing it will have some problems (either with an error or a panic) in the current implementation. In that case, could you add a failure test case if it's failing with an error, or add a check for this case if it panics for some reason?

Reply:
I'm also not sure if it even makes sense to apply From annotation to the variadic parameter. If it actually doesn't make sense, I think we should return an error when fromAnnotation.apply() is called for this case.

Reply:
i think it doesn't make sense..